runtime/src/main/java/org/corfudb/runtime/Restore.java

```diff
@@ -0,0 +1,174 @@
+ *
+ * Restore all tables in the given backup .tar file which is generated by Corfu Backup.
+ *
+ * Steps:
+ * 1. Open the given .tar file, obtain a set of backup files for tables
+ * 2. Restore tables by committing transactions using the OpaqueEntries from table backup files
+ */
+@Slf4j
+public class Restore {
+
+    // The path of backup tar file
+    private final String filePath;
+
+    // The path of a temporary directory under which the unpacked table's backup files are stored
+    private final String restoreTempDirPath;
+
+    // The stream IDs of tables which are restored
+    private List<UUID> streamIDs;
+
+    // The Corfu Store associated with the runtime
+    private CorfuStore corfuStore;
+
+    /**
+     * Unpacked files from backup tar file are stored under RESTORE_TEMP_DIR. They are deleted after restore finishes.
+     */
+    private static final String RESTORE_TEMP_DIR_PREFIX = "corfu_restore_";
+
+    /***
+     * @param filePath      - the path of backup tar file
+     * @param runtime       - the runtime which is performing the restore
+     * @throws IOException when failed to create the temp directory
+     */
+    public Restore(String filePath, CorfuRuntime runtime) throws IOException {
+        this.filePath = filePath;
+        this.restoreTempDirPath = Files.createTempDirectory(RESTORE_TEMP_DIR_PREFIX).toString();
+        this.streamIDs = new ArrayList<>();
+        this.corfuStore = new CorfuStore(runtime);
+    }
+
+    /**
+     * Start the restore process
+     *
+     * @throws IOException
+     */
+    public void start() throws IOException {
+        try {
+            openTarFile();
+            verify();
+            restore();
+        } catch (IOException e) {
+            log.error("failed to restore from backup tar file {}", filePath);
+            throw e;
+        } finally {
+            cleanup();
+        }
+        log.info("restore completed");
+    }
+
+    private void restore() throws IOException {
+        for (UUID streamId : streamIDs) {
+            String fileName = restoreTempDirPath + File.separator + streamId;
+            try {
+                restoreTable(fileName, streamId);
+            } catch (IOException e) {
+                log.error("failed to restore table {} from temp file {}", streamId, fileName);
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Restore a single table
+     *
+     * @param fileName   - the name of the temp backup file
+     * @param streamId   - the stream ID of the table which is to be restored
+     * @throws IOException
+     */
+    private void restoreTable(String fileName, UUID streamId) throws IOException {
+        try (FileInputStream fileInput = new FileInputStream(fileName)) {
+            // Clear table before restore
+            TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE);
+            SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
```

Comment:
should we also clear the stream and get rid of any ongoing updates?

Reply:
When this particular SMREntry gets committed it will clear the stream to empty. Also, the team using this Restore guarantees that all services are shut down during restore so there will be no more ongoing updates after the clear request.